doc/cli_wrapper/shd-cli_wrapper_hld.md

```diff
@@ -0,0 +1,213 @@
+
+* CLI_worker
+
+  Decode the request and execute this command. Return the execution result to client and finish this process.
+
+  Everytime cli_wapper server receives a request, it spawns a process to handle this request, so no single block point may happen in this design. 
+
+* CLI_wrapper client
+
+  It is a library that can be used to interact with CLI wrapper server. 
+
+---
+
+## 1.3 Design
+
+This section provides design details for CLI_wrapper.
+
+### 1.3.1 Generate wrapped files
+
+![Generate wrapper file](images/generate_wrapper_file.jpg)
+
+* In the step 1 and step 2, builder copies the necessary libraries to wrapped file and discard other libraries to reduce loading time
+
+* In the step 3, builder removes the content of command function body and replaces it with CLI_wrapper client
+
+### 1.3.2 Command execution flow
+
+![Command execution flow](images/command_execution_flow.jpg)
+
+* System ctrl bring up CLI_wrapper server when system starts
+
+* Click engine imports wrapped main.py and call CLI_wrapper client to send request
+
+* CLI_wrapper client gets command parameter and environment setting from CLI context, then encodes those information to a command string and send to CLI_wrapper server.
+
+* CLI_wrapper server launches a CLI worker to deal with this request and write timestamp and pid to local DB
+
+* CLI worker decodes the command to get parameters and environment setting, then execute this command based on those settings. In the end, it sends return code and data back to CLI wrapper client
+
+### 1.3.3 Watchdog execution flow
+
+![Command execution flow](images/watchdog_flow.jpg)
+* Default polling time is 60 seconds, watchdog check if the pid in local DB is still running, if not, update local DB.
+
+* Default expired time is 600 seconds (To send terminate signal) and 660 seconds (To send kill signal), watchdog check if CLI worker execution time exceeds expired time, send terminate or kill signal to recycle it.
+
+### 1.3.4 Encoded command format
+
+In order to send all needed information in a request, some encode processes must be applied before sending it to server.
+The following picture shows two examples for encoded requests.
+
+![Encode format](images/encode_format.jpg)
+
+* \__wrapper_calling_cmd__: used to indicate the start of a command
+
+* \__wrapper_env__ : used to indicate the environment setting when executing this command, separate the value by ";", this is a optional field.
+
+### 1.3.5 CLI wrapper service
+
+In compile stage, CLI_wrapper builder put generated wrapped file and original main.py to a backup folder.
+
+CLI wrapper service use these backup files to enable/disable CLI_wrapper feature.
+
+![Encode format](images/compiler_stage.jpg)
+
+* Enable CLI wrapper service
+
+![Encode format](images/wrapper_enable.jpg)
+
+    1. Replace main.py with wrapped main.py
+
+    2. Start CLI wrapper daemon
+
+* Disable CLI wrapper service
+
+![Encode format](images/wrapper_disable.jpg)

```

Comment:
How is this done?  Does a user have to manually rename the file and stop the server?  A global rules flag set to "n" should take care of these steps instead of manual intervention on every switch.